src/hotspot/cpu/riscv/stubGenerator_riscv.cpp

```diff
@@ -4809,6 +4809,348 @@ class StubGenerator: public StubCodeGenerator {
+      __ xorr(cur_w, cur_w, t1);
+
+      __ xorr(cur_w, cur_w, t0);
+      __ rolw_imm(cur_w, cur_w, 1, t0);
+
+      // copy the cur_w value to ws[8].
+      // now, valid w't values are at:
+      //  w0:       ws[0]'s lower 32 bits
+      //  w1 ~ w14: ws[1] ~ ws[7]
+      //  w15:      ws[8]'s higher 32 bits
+      __ slli(ws[idx/2], cur_w, 32);
+
+      return;
+    }
+
+    int idx = 17;
+    // W't = ROTL'1(W't-3 ^ W't-8 ^ W't-14 ^ W't-16),  16 <= t <= 79
+    __ srli(t0, ws[(idx-3)/2], 32);
+    __ mv(t1, ws[(idx-8)/2]);
+    __ xorr(t0, t0, t1);
+
+    __ mv(t1, ws[(idx-14)/2]);
+    __ mv(cur_w, ws[(idx-16)/2]);
+    __ xorr(cur_w, cur_w, t1);
+
+    __ xorr(cur_w, cur_w, t0);
+    __ rolw_imm(cur_w, cur_w, 1, t0);
+
+    // copy the cur_w value to ws[8]
+    __ zero_extend(cur_w, cur_w, 32);
+    __ orr(ws[idx/2], ws[idx/2], cur_w);
+
+    // shift the w't registers, so they start from ws[0] again.
+    // now, valid w't values are at:
+    //  w0 ~ w15: ws[0] ~ ws[7]
+    Register ws_0 = ws[0];
+    for (int i = 0; i < 16/2; i++) {
+      ws[i] = ws[i+1];
+    }
+    ws[8] = ws_0;
+  }
+
+  // f't(x, y, z) =
+  //    Ch(x, y, z)     = (x & y) ^ (~x & z)            , 0  <= t <= 19
+  //    Parity(x, y, z) = x ^ y ^ z                     , 20 <= t <= 39
+  //    Maj(x, y, z)    = (x & y) ^ (x & z) ^ (y & z)   , 40 <= t <= 59
+  //    Parity(x, y, z) = x ^ y ^ z                     , 60 <= t <= 79
+  void sha1_f(int round, Register dst, Register x, Register y, Register z) {
+    assert(round >= 0 && round < 80, "must be");
+    assert_different_registers(dst, x, y, z);
+
+    if (round < 20) {
+      // (x & y) ^ (~x & z)
+      __ andr(t0, x, y);
+      __ andn(dst, z, x);
+      __ xorr(dst, dst, t0);
+    } else if (round >= 40 && round < 60) {
+      // (x & y) ^ (x & z) ^ (y & z)
+      __ andr(t0, x, y);
+      __ andr(t1, x, z);
+      __ andr(dst, y, z);
+      __ xorr(dst, dst, t0);
+      __ xorr(dst, dst, t1);
+    } else {
+      // x ^ y ^ z
+      __ xorr(dst, x, y);
+      __ xorr(dst, dst, z);
+    }
+  }
+
+  // T = ROTL'5(a) + f't(b, c, d) + e + K't + W't
+  // e = d
+  // d = c
+  // c = ROTL'30(b)
+  // b = a
+  // a = T
+  void sha1_process_round(int round, Register a, Register b, Register c, Register d, Register e,
+                          Register cur_k, Register cur_w, Register tmp) {
+    assert(round >= 0 && round < 80, "must be");
+    assert_different_registers(a, b, c, d, e, cur_w, cur_k, tmp);
+
+    // T = ROTL'5(a) + f't(b, c, d) + e + K't + W't
+    {
+      // reuse e as a temporary register, as we will mv new value into it later
+      Register t = e;
+      // cur_w will be recalculated at the beginning of each round,
+      // so, we can reuse it as a temp register here.
+      __ add(cur_w, cur_k, cur_w);
+      __ add(t, t, cur_w);
+      __ rolw_imm(cur_w, a, 5, t0);
+      // as pointed above, we can use cur_w as temporary register here.
+      sha1_f(round, tmp, b, c, d);
```

Comment:
Seems not, as cur_k is only calculated when `(round % 20) == 0`.

Other comments are all resolved.

Reply:
Could you please move `int round` as the last param for other functions so that it will be more consistent? They are `sha1_prepare_w`, `sha1_f`, and `sha1_process_round`. BTW: I think it will be safer to add scratch register `t0`, `t1` to the list for `assert_different_registers` if they are used in those assembler functions.

Reply:
Sure, it's updated. Thanks.